File: core/src/validators/jsonld.ts
// core/src/validators/jsonld.ts
//
// JSON-LD validator orchestrator. Parses input, delegates per-item
// validation to per-item.ts. Returns a ValidationResult with
// format="jsonld" and the loaded registry's provenance fields.
//
// Strictly synchronous, in-memory, single pass over the document.
// Honors constitution Design Tenet #3.

import type {
  Issue,
  ValidationResult,
  ValidateOptions,
} from "../types.js";
import { loadRegistry } from "../registry.js";
import { loadCuratedRules } from "../curated-rules.js";
import { parseError } from "../errors.js";
import { validateItem } from "./per-item.js";

const registry = loadRegistry();
const curatedRules = loadCuratedRules();

function bucket(issues: Issue[]): {
  errors: Issue[];
  warnings: Issue[];
  info: Issue[];
} {
  const errors: Issue[] = [];
  const warnings: Issue[] = [];
  const info: Issue[] = [];
  for (const issue of issues) {
    if (issue.type === "error") errors.push(issue);
    else if (issue.type === "warning") warnings.push(issue);
    else info.push(issue);
  }
  return { errors, warnings, info };
}

function registryProvenance() {
  return {
    schemaVersion: registry.schemaVersion,
    snapshotAt: registry.snapshotAt,
    curatedRulesVersion: curatedRules.sourceVersion,
  };
}

export function validateJsonLd(
  input: unknown,
  options: ValidateOptions = {},
): ValidationResult {
  // 1) Parse if string.
  let obj: unknown;
  if (typeof input === "string") {
    try {
      obj = JSON.parse(input);
    } catch (err) {
      const reason = err instanceof Error ? err.message : String(err);
      return {
        valid: false,
        format: "jsonld",
        types: [],
        errors: [parseError(input, reason)],
        warnings: [],
        info: [],
        registry: registryProvenance(),
      };
    }
  } else {
    obj = input;
  }

  // 2) Ensure it's an object.
  if (obj === null || typeof obj !== "object" || Array.isArray(obj)) {
    return {
      valid: false,
      format: "jsonld",
      types: [],
      errors: [parseError(input, "Input is not a JSON object")],
      warnings: [],
      info: [],
      registry: registryProvenance(),
    };
  }

  // 3) Validate via the shared per-item engine.
  const { issues, resolvedType } = validateItem(
    obj as Record<string, unknown>,
  );

  // 4) Bucket + strict.
  const { errors, warnings, info } = bucket(issues);
  const strictFail = options.strict === true && warnings.length > 0;

  return {
    valid: errors.length === 0 && !strictFail,
    format: "jsonld",
    types: resolvedType ? [resolvedType] : [],
    errors,
    warnings,
    info,
    registry: registryProvenance(),
  };
}

File: python/src/schema_audit/validators/jsonld.py
"""JSON-LD orchestrator.

Parses input (if string), routes the parsed object through the shared
per-item engine, buckets issues by severity, applies ``strict``, and
returns a fully-populated :class:`schema_audit.types.ValidationResult`.

Strictly synchronous, single-pass, no I/O. Mirrors
``core/src/validators/jsonld.ts``.
"""

from __future__ import annotations

import json
from json import JSONDecodeError

from .._per_item import validate_item
from ..curated_rules import load_curated_rules
from ..errors import parse_error
from ..registry import load_registry
from ..types import Issue, RegistryInfo, ValidationResult

_REGISTRY = load_registry()
_CURATED = load_curated_rules()


def _registry_provenance() -> RegistryInfo:
    return {
        "schemaVersion": _REGISTRY["schemaVersion"],
        "snapshotAt": _REGISTRY["snapshotAt"],
        "curatedRulesVersion": _CURATED["sourceVersion"],
    }


def _bucket(issues: list[Issue]) -> tuple[list[Issue], list[Issue], list[Issue]]:
    errors: list[Issue] = []
    warnings: list[Issue] = []
    info: list[Issue] = []
    for issue in issues:
        severity = issue["type"]
        if severity == "error":
            errors.append(issue)
        elif severity == "warning":
            warnings.append(issue)
        else:
            info.append(issue)
    return errors, warnings, info


def _parse_error_result(original_input: object, reason: str) -> ValidationResult:
    return {
        "valid": False,
        "format": "jsonld",
        "types": [],
        "errors": [parse_error(original_input, reason)],
        "warnings": [],
        "info": [],
        "registry": _registry_provenance(),
    }


def validate_jsonld(input_value: object, *, strict: bool = False) -> ValidationResult:
    """Validate JSON-LD input (string, dict, or list).

    The behavior matches the JS orchestrator exactly:

    - string input is JSON-parsed; ``JSONDecodeError`` → ``PARSE_ERROR``
    - top-level list / None / non-dict → ``PARSE_ERROR``
    - dict input → routed through the per-item engine
    """
    # 1. Parse if string.
    obj: object
    if isinstance(input_value, str):
        try:
            obj = json.loads(input_value)
        except JSONDecodeError as err:
            return _parse_error_result(input_value, str(err))
    else:
        obj = input_value

    # 2. Top-level shape must be a dict.
    if not isinstance(obj, dict):
        return _parse_error_result(input_value, "Input is not a JSON object")

    # 3. Validate via shared per-item engine.
    result = validate_item(obj)

    # 4. Bucket + strict.
    errors, warnings, info = _bucket(result.issues)
    strict_fail = strict and len(warnings) > 0

    return {
        "valid": len(errors) == 0 and not strict_fail,
        "format": "jsonld",
        "types": [result.resolved_type] if result.resolved_type else [],
        "errors": errors,
        "warnings": warnings,
        "info": info,
        "registry": _registry_provenance(),
    }
